R/derive_merged.R

```diff
@@ -0,0 +1,685 @@
+#'   respect to the specified order.
+#'
+#'   *Permitted Values*: list of variables or `desc(<variable>)` function calls
+#'
+#' @param new_vars Variables to add
+#'
+#'   The specified variables are added to the output dataset. Variables can be
+#'   renamed by naming the element, i.e., `new_vars = vars(<new name> = <old
+#'   name>)`.
+#'
+#'   If the parameter is not specified or set to `NULL`, all variables from the
+#'   additional dataset are added.
+#'
+#'   *Permitted Values*: list of variables created by `vars()`
+#'
+#' @param mode Selection mode
+#'
+#'   Determines if the first or last observation is selected.
+#'
+#'   *Permitted Values*: `"first"`, `"last"`
+#'
+#' @param by_vars Grouping variables
+#'
+#'   The input dataset and the selected observations from the additional dataset
+#'   are merged by the specified by variables. The by variables must be a unique
+#'   key of the selected observations.
+#'
+#'   *Permitted Values*: list of variables
+#'
+#' @param filter_add Filter for additional data
+#'
+#'   Only observations fulfilling the specified condition are taken into account
+#'   for merging. If the parameter is not specified, all observations are
+#'   considered.
+#'
+#'   *Permitted Values*: a condition
+#'
+#' @param match_flag Match flag
+#'
+#'   If the parameter is specified, the specified variable is added to the
+#'   output dataset and set to `TRUE` for all by groups with an observation in
+#'   the additional dataset. For all other by groups it is set to `NA`.
+#'
+#'   *Permitted Values*: Variable name
+#'
+#' @param check_type Check uniqueness?
+#'
+#'   If `"warning"` or `"error"` is specified, the specified message is issued
+#'   if the observations of the (restricted) additional dataset are not unique
+#'   with respect to the by variables and the order.
+#'
+#'   *Permitted Values*: `"none"`, `"warning"`, `"error"`
+#'
+#' @details
+#'
+#'   1. The additional dataset is restricted to the observations matching the
+#'   `filter_add` condition.
+#'
+#'   1. If `order` is specified, for each by group the first or last observation
+#'   (depending on `mode`) is selected.
+#'
+#'   1. The variables specified for `new_vars` are merged to the input dataset.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @keywords derivation adam
+#'
+#' @export
+#'
+#' @examples
+#' library(admiral.test)
+#' library(dplyr, warn.conflicts = FALSE)
+#' data("vs")
+#' data("dm")
+#'
+#' # merging all dm variables to vs
+#' derive_vars_merged(
+#'   vs,
+#'   dataset_add = select(dm, -DOMAIN),
+#'   by_vars = vars(STUDYID, USUBJID)) %>%
+#' select(STUDYID, USUBJID, VSTESTCD, VISIT, VSTPT, VSSTRESN, AGE, AGEU)
+#'
+#' # merge last weight to adsl
+#' data("adsl")
+#' derive_vars_merged(
+#'   adsl,
+#'   dataset_add = vs,
+#'   by_vars = vars(STUDYID, USUBJID),
+#'   order = vars(VSDTC),
+#'   mode = "last",
+#'   new_vars = vars(LASTWGT = VSSTRESN, LASTWGTU = VSSTRESU),
+#'   filter_add = VSTESTCD == "WEIGHT"
```

Comment:
For the merge functions the selection is based on `dataset_add` only.

I agree that we should have a function for your use case. We have already an issue for it (https://github.com/pharmaverse/admiral/issues/578).

I would not integrate this functionality into the merge functions because the flow is different and it requires much more resources. The datasets need to be joined first and then restricted. Joining first can lead to huge datasets. Thus it should be done only if it is really required.